File: Java/leitorOfx.java
public class OfxReader {

    public static void main(String[] args) {
        String path = "/home/rodrigovidal/Documentos/materiais_de_estudo/extrato.ofx";
        try {
            InputStream entrada = new FileInputStream(path);
            new OfxReader().lerOfx(entrada);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    public void lerOfx(InputStream entrada) {
        AggregateUnmarshaller<ResponseEnvelope> a = new AggregateUnmarshaller<>(ResponseEnvelope.class);

        try {
            ResponseEnvelope re = a.unmarshal(entrada);
            //SignonResponse sr = re.getSignonResponse();
            MessageSetType type = MessageSetType.banking;
            ResponseMessageSet message = re.getMessageSet(type);

            if (message != null) {
                List<BankStatementResponseTransaction> bank = ((BankingResponseMessageSet) message).getStatementResponses();

                for (BankStatementResponseTransaction b : bank) {
                    System.out.println("cc: " + b.getMessage().getAccount().getAccountNumber());
                    System.out.println("ag: " + b.getMessage().getAccount().getBranchId());
                    System.out.println("balanço final: " + b.getMessage().getLedgerBalance().getAmount());
                    System.out.println("dataDoArquivo: " + b.getMessage().getLedgerBalance().getAsOfDate());
                    List<Transaction> list = b.getMessage().getTransactionList().getTransactions();
                    System.out.println("TRANSAÇÕES\n");
                    for (Transaction transaction : list) {
                        System.out.println("tipo: " + transaction.getTransactionType().name());
                        System.out.println("id: " + transaction.getId());
                        System.out.println("data: " + transaction.getDatePosted());
                        System.out.println("valor: " + transaction.getAmount());
                        System.out.println("descricao: " + transaction.getMemo() + "\n");
                    }
                }
            }
        } catch (IOException | OFXParseException e) {
            e.printStackTrace();
        }
    }

}

//DEPENDENCIAS
// https://mvnrepository.com/artifact/org.javassist/javassist            3.26.0-GA
// https://mvnrepository.com/artifact/org.reflections/reflections        0.9.11
// https://mvnrepository.com/artifact/commons-logging/commons-logging    1.2
// https://mvnrepository.com/artifact/be.cyberelf.nanoxml/nanoxml        2.2.3
// https://mvnrepository.com/artifact/com.webcohesion.ofx4j/ofx4j        1.17
// https://mvnrepository.com/artifact/com.google.guava/guava             28.1-jre

File: Kotlin/LeitorOfx.kt
package utils

import com.webcohesion.ofx4j.domain.data.MessageSetType
import com.webcohesion.ofx4j.domain.data.ResponseEnvelope
import com.webcohesion.ofx4j.domain.data.ResponseMessageSet
import com.webcohesion.ofx4j.domain.data.banking.BankStatementResponseTransaction
import com.webcohesion.ofx4j.domain.data.banking.BankingResponseMessageSet
import com.webcohesion.ofx4j.io.AggregateUnmarshaller
import com.webcohesion.ofx4j.io.OFXParseException
import java.io.Reader
import java.io.StringReader
import java.util.*

/**
 * Classe LeitorOfx - Responsável por interpretar o arquivo OFX.
 * bank.message.transactionList.transactions
 * implementation group: 'com.webcohesion.ofx4j', name: 'ofx4j', version: '1.24'
 * */
class LeitorOfx {

    /**
     * Recebe o conteudo da requisição.
     *
     * @param content String - Conteúdo em base64
     * @return MutableList<BankStatementResponseTransaction>?
     * @throws OFXParseException
     * */
    @Throws(OFXParseException::class)
    fun recebeConteudo(content: String): MutableList<BankStatementResponseTransaction> {
        val conteudo = String(Base64.getDecoder().decode(content))
        val reader: Reader = StringReader(conteudo)
        return leDados(reader)
    }

    /**
     * Converte o OFX em um {@see BankStatementResponseTransaction}.
     *
     * @param entrada Reader
     * @return MutableList<BankStatementResponseTransaction>?
     * @throws OFXParseException
     * */
    @Throws(OFXParseException::class)
    private fun leDados(entrada: Reader): MutableList<BankStatementResponseTransaction> {

        val aggregate: AggregateUnmarshaller<ResponseEnvelope> = AggregateUnmarshaller(ResponseEnvelope::class.java)

        val re: ResponseEnvelope = aggregate.unmarshal(entrada)
        val type: MessageSetType = MessageSetType.banking
        val message: ResponseMessageSet? = re.getMessageSet(type)

        return if (message != null) {
            (message as BankingResponseMessageSet).statementResponses
        } else {
            throw OFXParseException("Erro ao ler arquivo.")
        }

    }
}
